src/Functions/FunctionsConversion.h

```diff
@@ -1874,7 +1882,11 @@ class FunctionConvertFromString : public IFunction
         }
 
         if constexpr (exception_mode == ConvertFromStringExceptionMode::Null)
-            res = std::make_shared<DataTypeNullable>(res);
+        {
+            auto to_type = WhichDataType(res);
+            if (!(isInteger(arguments[0].type) && (to_type.isDateOrDate32() || to_type.isDateTime() || to_type.isDateTime64())))
```

Comment:
This class is still called `FunctionConvertFromString`, but after this changes it allows know to parse also integers for date data types and there is `ConvertFromStringExceptionMode`, which is used not only for strings, but its name tells about strings only.

May be better to rename it to functionConvertWIthExceptionMode and mode enum to convertExceptionMode since it is easily adjusted not only for strings and the purpose of this function is to handle conversion with exception mode.